File: Bit Manipulation/reverse_bit.py
'''
Difficulty: Easy

https://leetcode.com/problems/reverse-bits/

Reverse bits of a given 32 bits unsigned integer.

Note:

Note that in some languages, such as Java, there is no unsigned integer type. In this case, both input and output will be given as a signed integer type. They should not affect your implementation, as the integer's internal binary representation is the same, whether it is signed or unsigned.
In Java, the compiler represents the signed integers using 2's complement notation. Therefore, in Example 2 above, the input represents the signed integer -3 and the output represents the signed integer -1073741825.
 
'''


class Solution:
    def reverseBits(self, n: int) -> int:
        result = 0

        for i in range(32):
            bit = (n >> i) & 1
            result = result | (bit << (31 - i))

        return result

File: 00.Rust/reverse_bit.rs
impl Solution {
    pub fn reverse_bits(x: u32) -> u32 {
        let mut result = 0;
        for i in 0..32{
            let bit = x & 1 << i;
            let travel = (31 - i as i32) - i as i32;
            if travel > 0{
                result |= bit << travel;
            } 
            else{
                result |= bit >> -travel;
            }

            }
            result
        }    
    }
